pkg/api/validate/quota.go

```diff
@@ -0,0 +1,93 @@
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (dv *openShiftClusterDynamicValidator) validateQuotas(ctx context.Context, oc *api.OpenShiftCluster) error {
+	//prepare usage client
+	spp := &oc.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(oc.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return dv.calculateAndCompareQuota(ctx, oc, usageClient)
+}
+
+func (dv *openShiftClusterDynamicValidator) addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, count int) error {
+	requiredResources["virtualMachines"] += count
+	requiredResources["PremiumDiskCount"] += count
+	switch vmSize {
+	case api.VMSizeStandardD2sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 2)
+	case api.VMSizeStandardD4sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 4)
+	case api.VMSizeStandardD8sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 8)
+	default:
+		//will only happen if pkg/api verification allows new VMSizes
+		return fmt.Errorf("Unexpected node VMSize %s.", vmSize)
+	}
+	return nil
+}
+
+func (dv *openShiftClusterDynamicValidator) calculateAndCompareQuota(ctx context.Context, oc *api.OpenShiftCluster, uc compute.UsageClient) error {
+	requiredResources := map[string]int{}
+	dv.addRequiredResources(requiredResources, oc.Properties.MasterProfile.VMSize, 3)
+	//worker node resource calculation
+	for _, w := range oc.Properties.WorkerProfiles {
+		dv.addRequiredResources(requiredResources, w.VMSize, w.Count)
+	}
+
+	usages, err := uc.List(ctx, oc.Location)
+	if err != nil {
+		return err
+	}
+	//check requirements vs. usage
+	var resources []string
+	for r := range requiredResources {
+		resources = append(resources, r)
+	}
+	sort.Strings(resources)
+
+	for _, r := range resources {
+		var usage *mgmtcompute.Usage
+		for _, u := range usages {
+			if *u.Name.Value == r {
+				usage = &u
+				break
+			}
+		}
+		if usage == nil {
+			return fmt.Errorf("Resource %s not in usage results.", r)
+		}
+		required, _ := requiredResources[r]
+		if int64(required) > (*usage.Limit - int64(*usage.CurrentValue)) {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeResourceQuotaExceeded, ".", "Insufficient '%s' quota: need %d, current available is %d (%d of %d used).", *usage.Name.Value, required, *usage.Limit-int64(*usage.CurrentValue), *usage.CurrentValue, *usage.Limit)
```

Comment:
Still trying to fine-tune this.

`return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeResourceQuotaExceeded, "", "Resource quota of %s exceeded.  Maximum allowed: %d, Current in use: %d, Additional requested: %d.", *usage.Name.Value, *usage.CurrentValue, *usage.Limit, required)`